interop-scoring/main.js

```diff
@@ -0,0 +1,506 @@
+  '/webcodecs/full-cycle-test.https.any.worker.html?vp8',
+  // interop-2023-webcomponents
+  '/shadow-dom/focus/focus-shadowhost-display-none.html',
+  '/custom-elements/form-associated/ElementInternals-labels.html',
+  '/custom-elements/form-associated/ElementInternals-setFormValue.html',
+  '/custom-elements/form-associated/ElementInternals-validation.html',
+  '/custom-elements/form-associated/form-disabled-callback.html',
+]);
+
+
+// Calculate interop score (passing in all browsers) for a category
+// after tracking the category's scores for each browser.
+function aggregateInteropTestScores(interopScores, numBrowsers) {
+  if (interopScores.length === 0) return 0;
+  let aggregateScore = 0;
+  for (const [, results] of interopScores) {
+    let subtestsAllPassing = 0;
+    for (const [, subtestNumPassing] of results) {
+      // A subtest counts toward the interop score if every browser passed the subtest.
+      // In other words, the number of passes should match the number of browsers.
+      if (subtestNumPassing === numBrowsers) {
+        subtestsAllPassing += 1;
+      }
+    }
+    aggregateScore += Math.floor(1000 * subtestsAllPassing / results.size);
+  }
+  return Math.floor(aggregateScore / interopScores.size) || 0;
+}
+
+// Score a set of runs (independently) on a set of tests. The runs are presumed
+// to be aligned in some way (i.e. they were all run at the same WPT SHA).
+//
+// Returns an array of scores, which is the top-level score (integer 0-1000) for
+// each corresponding input run.
+//
+// To get the top-level score for a run, each test in that run that is present
+// in |allTestsSet| is examined. Each test is scored 0-1000 based on the
+// fraction of its subtests that pass, with rounding down so that 1000 means
+// all subtests pass. Reftests score either 0 or 1000. These test scores are
+// then summed and divided by the size of |allTestsSet|, again rounding down.
+//
+// This methodology has several consequences:
+//
+//   1. Individual tests do have a heavier weight than subtests. This could be
+//   gamed, by splitting passing tests into multiple files rather than using
+//   subtests (or conversely by combining failing tests into subtests in a
+//   single file).
+//
+//   2. If |allTestsSet| is constant across runs *through time*, older runs may
+//   not have entries for tests were only added recently and will be penalized
+//   for that. This is deliberate - see the comment block later in this
+//   function for why.
+//
+//   3. We could show (on wpt.fyi) scores at both the test and category level as
+//   a percentage with one decimal point, and what a user would see would be the
+//   same numbers that go into the total score, with no hidden rounding error.
+//
+//   4. Because we round down twice, the score for a category can end up lower
+//   than if we used rational numbers.
+function scoreRuns(runs, allTestsSet) {
+  const scores = [];
+  const interopScores = new Map();
```

Comment:
Any part of this function related to the `interopScores` Map is newly added to capture all test/subtest information to be aggregated later using the `aggregateInteropTestScores` function.

Reply:
Do you have any idea for another name for this data structure? It isn't the interop score yet, it's the input to compute the interop score in `aggregateInteropTestScores`. If I'm reading everything right, it's the number of passes per subtest. Maybe `passCounts` or `testPassCounts`? Or `aggregateSomething`?

Reply:
That's true. I've changed the name to `testPassCounts` like you suggested.